File: Fore-C_v2/codes/Grid_covariates_sst_metrics.R
# Pull SST covariates to grid --------------------------------------------------

# load libraries
library(ncdf4)
library(tidyverse)
library(httr)

# source custom function 
source("./codes/custom_functions/fun_ftp_download.R")

# list weekly SST files -----------------------------------------------

parent_ftp_filepath <- 'ftp://ftp.star.nesdis.noaa.gov/pub/sod/mecb/crw/data/for_forec/shiny_app/'

# list directories, named by date
files <- list_ftp_files(ftp_path = parent_ftp_filepath)

parent_file <- max(as.numeric(files), na.rm = T)

child_ftp_filepath <- paste0(parent_ftp_filepath, parent_file, '/')

files <- list_ftp_files(ftp_path = child_ftp_filepath)

# load and format data ------------------------------------------------

# create empty data frame
sst_metrics <- data.frame()

# open and format data (this should take 1-4 minutes on a standard laptop)
for(j in files){
  # open data
  ftp_path_tmp <- paste0(child_ftp_filepath, j)
  res <- GET(ftp_path_tmp, write_disk(basename(ftp_path_tmp), overwrite = TRUE))
  x <- nc_open(res$request$output$path)
  # get ids
  ids <- ncvar_get(x, varid = "reef_id")
  # get date(s)
  dates <- ncvar_get(x, varid = "data_date")
  # determine metric name
  if(grepl('hotsnap|hdw', j) == TRUE){
    metric_name <- 'Hot_snaps'
  } else if(grepl('mean-90d', j) == TRUE){
    metric_name <- 'SST_90dMean'
  } else {
    metric_name <- 'Winter_condition'
  }
  # determine if each file is a cfs forecast (True) or satellite measurement (False) 
  cfs <- ifelse(grepl('cfsv2', j) == TRUE, TRUE, FALSE)
  # format forecasts
  if(cfs == TRUE){
    metric_array <- ncvar_get(x, varid = names(x$var)[1])
    tmp_df <- data.frame()
    for(k in 1:12){ # dates
      for(l in 1:28){ # ensembles
        tmp_metric <- metric_array[, l, k]
        # make flagged values NAs
        tmp_metric[tmp_metric == 253|tmp_metric == 9999] <- NA
        # fill in NAs with median values
        tmp_metric[is.na(tmp_metric)] <- median(tmp_metric, na.rm = T)
        # create dataframe
        cfs_df <- data.frame("ID" = ids
                             , "Date" = dates[k]
                             , "ensemble" = l
                             , "type" = "forecast"
                             , "temp_metric_name" = metric_name
                             , "value" = tmp_metric
                             )
        tmp_df <- rbind(tmp_df, cfs_df)
        }
      }
    # format sst measurements
    } else if (cfs == FALSE){
    tmp_metric <- ncvar_get(x, varid = names(x$var)[1])
    # make flagged values NAs
    tmp_metric[tmp_metric == 253|tmp_metric == 9999] <- NA
    # fill in NAs with median values
    tmp_metric[is.na(tmp_metric)] <- median(tmp_metric, na.rm = T)
    # create dataframe
    tmp_df <- data.frame("ID" = ids
                         , "Date" = dates
                         , "ensemble" = 0
                         , "type" = "nowcast"
                         , "temp_metric_name" = metric_name
                         , "value" = tmp_metric
                         )
    }
  # combine data
  sst_metrics <- rbind(sst_metrics
                       , tmp_df)
  nc_close(x)
  file.remove(j)
  cat("finished ", j, '\n')
}

# files <- files[which(files == j):length(files)]

# reshape files ----------------------------------------------
reef_grid_sst <- sst_metrics %>%
  spread(key = temp_metric_name, value = value) %>%
  fill(c("Hot_snaps", "SST_90dMean" , "Winter_condition"), .direction = 'updown')

# format dates
reef_grid_sst$Date <- as.Date(reef_grid_sst$Date, "%Y-%m-%d")

# Add regions and offset wdw here
# add region, then update winter condition based on region
load('../compiled_data/spatial_data/grid.RData')
reef_grid_sst <- reef_grid_sst %>%
  left_join(reefsDF)

reef_grid_sst$Region[reef_grid_sst$Longitude >= 140 & reef_grid_sst$Longitude <= 155 & reef_grid_sst$Latitude >= -28 & reef_grid_sst$Latitude <= -7] <- "gbr"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -180 & reef_grid_sst$Longitude <= -152 & reef_grid_sst$Latitude >= 18 & reef_grid_sst$Latitude <= 30] <- "hawaii"
reef_grid_sst$Region[reef_grid_sst$Longitude >= 143 & reef_grid_sst$Longitude <= 147 & reef_grid_sst$Latitude >= 12 & reef_grid_sst$Latitude <= 21] <- "guam-cnmi"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -174 & reef_grid_sst$Longitude <= -167 & reef_grid_sst$Latitude >= -16 & reef_grid_sst$Latitude <= -10] <- "samoas"
reef_grid_sst$Region[reef_grid_sst$Longitude >= 165 & reef_grid_sst$Longitude <= 168 & reef_grid_sst$Latitude >= 18 & reef_grid_sst$Latitude <= 21] <- "wake"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -171 & reef_grid_sst$Longitude <= -168 & reef_grid_sst$Latitude >= 15 & reef_grid_sst$Latitude <= 18] <- "johnston"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -178 & reef_grid_sst$Longitude <= -175 & reef_grid_sst$Latitude >= -1 & reef_grid_sst$Latitude <= 2] <- "howland-baker"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -161 & reef_grid_sst$Longitude <= -159 & reef_grid_sst$Latitude >= -2 & reef_grid_sst$Latitude <= 1] <- "jarvis"
reef_grid_sst$Region[reef_grid_sst$Longitude >= -164 & reef_grid_sst$Longitude <= -161 & reef_grid_sst$Latitude >= 4 & reef_grid_sst$Latitude <= 8] <- "palmyra-kingman"

reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'guam-cnmi', reef_grid_sst$Winter_condition - 3.73, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'howland-baker', reef_grid_sst$Winter_condition - 19.25, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'johnston', reef_grid_sst$Winter_condition - 2.85, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'samoas', reef_grid_sst$Winter_condition - 4.00, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'gbr', reef_grid_sst$Winter_condition - 1.95, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'hawaii', reef_grid_sst$Winter_condition - 2.73, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'jarvis', reef_grid_sst$Winter_condition - 18.96, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'palmyra-kingman', reef_grid_sst$Winter_condition - 7.01, reef_grid_sst$Winter_condition)
reef_grid_sst$Winter_condition <- ifelse(reef_grid_sst$Region == 'wake', reef_grid_sst$Winter_condition - 4.01, reef_grid_sst$Winter_condition)

# remove Region
reef_grid_sst$Region <- NULL

# save data --------------------------------------------------
save(reef_grid_sst, 
     file = "../compiled_data/grid_covariate_data/grid_with_sst_metrics.RData")

File: Fore-C_v2/codes/Grid_covariates_sst_metrics.py
# -*- coding: utf-8 -*-
"""
Created on Thu Dec  9 14:06:21 2021

@author: jamie
"""
import wget
import os
import netCDF4 as nc
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

# source custom functions
from codes.custom_functions.fun_ftp_download import list_ftp_files
from codes.custom_functions.fun_winter_condition_offset import winter_condition_offset

# list weekly SST files -----------------------------------------------

# list directories, named by date
parent_ftp_filepath = 'https://www.star.nesdis.noaa.gov/pub/sod/mecb/crw/data/for_forec/shiny_app/'

files = list_ftp_files(parent_ftp_filepath)

parent_file = [i for i in files if i.startswith('2')][-1]

child_ftp_filepath = parent_ftp_filepath + parent_file
 
files = list_ftp_files(child_ftp_filepath)

# remove extraneous file names
files = [i for i in files if i.startswith('cfs')|i.startswith('dz')|i.startswith('forec')]

# download data ---------------------------------------------
path = '../crw_weekly_data/'

# Check whether the specified path exists or not
isExist = os.path.exists(path)

# Create a new directory because it does not exist 
if not isExist:
  os.makedirs(path)

# download nc files
for jj in files:
    # download file
    nc_source_filepath = child_ftp_filepath + jj
    nc_dest_filepath = path + jj
    wget.download(nc_source_filepath, out = nc_dest_filepath)

# format data ------------------------------------------------

# create empty data frame
sst_metrics = pd.DataFrame()

# open and format data (this should take 1-4 minutes on a standard laptop)
for j in files:
    # open file
    x = nc.Dataset(path + j)
    # get ids
    ids = x.variables['reef_id'][:]
    ids = pd.DataFrame(ids, columns = ['ID'])
    # get date(s)
    dates = x.variables['data_date'][:]
    dates = pd.DataFrame(dates, columns = ['data_date'])
    # determine metric name
    if 'hotsnap' in j or 'hdw' in j:
        metric_name = 'Hot_snaps'
    elif 'mean-90d' in j: 
         metric_name = 'SST_90dMean'
    else: 
         metric_name = 'Winter_condition'
    # set variable id
    variable_dict = x.variables.keys()
    var_name = list(variable_dict)[2]
    metric_array = x.variables[var_name][:]
    metric_array = np.array(metric_array)
    tmp_df = pd.DataFrame()
    # format forecasts
    if 'cfs' in j:
        for k in range(12): # dates for first 12 weeks of forecasts
            for l in range(28): # ensembles
                # dimensions = dates x ids x values
                tmp_metric = metric_array[k, l, :]
                # make flagged values NaNs
                tmp_metric = np.where(tmp_metric == 253, np.nan, tmp_metric)
                tmp_metric = np.where(tmp_metric == 9999, np.nan, tmp_metric)
                # fill in NaNs with median values
                inds = np.where(np.isnan(tmp_metric))
                tmp_metric[inds] = np.nanmedian(tmp_metric)
                # create dataframe
                cfs_df = ids.copy(deep = True)
                data_date = np.array(dates.iloc[k].repeat(cfs_df.shape[0]))
                cfs_df['Date'] = data_date
                cfs_df['ensemble'] = l + 1
                cfs_df['type'] = 'forecast'
                cfs_df['temp_metric_name'] = metric_name
                cfs_df['value'] = tmp_metric
                tmp_df = tmp_df.append(cfs_df, ignore_index = True)
    else:
      tmp_metric = metric_array.copy() 
      # make flagged values NaNs
      tmp_metric = np.where(tmp_metric== 253, np.nan, tmp_metric)
      tmp_metric = np.where(tmp_metric== 9999, np.nan, tmp_metric)
      # fill in NaNs with median values
      inds = np.where(np.isnan(tmp_metric))
      tmp_metric[inds] = np.nanmedian(tmp_metric)
      # create dataframe
      tmp_df = ids.copy(deep = True)
      data_date = np.array(dates.loc[0].repeat(tmp_df.shape[0]))
      tmp_df['Date'] = data_date
      tmp_df['ensemble'] = 0
      tmp_df['type'] = 'nowcast'
      tmp_df['temp_metric_name'] = metric_name
      tmp_df['value'] = tmp_metric[0,]
    # combine data
    sst_metrics = sst_metrics.append(tmp_df)
    # close nc file
    x.close()
    os.remove(path + j)
    # print progress
    print('finished', j)

# delete directory of weekly CRW files
# for m in files:
#     os.remove(path + m)
os.rmdir(path)

# sst_metrics.to_csv('../compiled_data/grid_covariate_data/grid_with_sst_metrics.csv', index=False)

# reshape file ----------------------------------------------

# go from long to wide format
reef_grid_sst = sst_metrics.pivot_table(index = ['ID', 'Date', 'ensemble', 'type']
                                          , columns = 'temp_metric_name'
                                          , values = 'value').reset_index()

# Use near-real time Winter Condition for all forecast dates
# There is only one near-real time date in each update, so we have used mean - 
# If there are multiple near-real time dates, update code to use most recent date
# this can also be removed once forecasted WDW are available
reef_grid_sst['Winter_condition'] = reef_grid_sst.groupby('ID').transform(lambda x: x.fillna(x.mean()))

# save
reef_grid_sst.to_csv('../compiled_data/grid_covariate_data/grid_with_sst_metrics.csv', index = False)
# reef_grid_sst = pd.read_csv('../compiled_data/grid_covariate_data/grid_with_sst_metrics.csv')

# load reef grid
reefsDF = pd.read_csv('../compiled_data/spatial_data/grid.csv')

# merge sst data with reef grid
reef_grid_sst = reef_grid_sst.merge(reefsDF, on = 'ID', how = 'left')

# update Winter Condition based on region
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'guam-cnmi', offset_value = 3.73)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'howland-baker', offset_value = 19.25)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'johnston', offset_value = 2.85)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'samoas', offset_value = 4.00)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'gbr', offset_value = 1.95)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'hawaii', offset_value = 2.73)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'jarvis', offset_value = 18.96)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'palmyra-kingman', offset_value = 7.01)
winter_condition_offset(df = reef_grid_sst, crw_vs_region_name = 'wake', offset_value = 4.01)

# save data --------------------------------------------------
reef_grid_sst.to_csv('../compiled_data/grid_covariate_data/grid_with_sst_metrics.csv', index = False)

